go/protomodule/protomodule_message.go

```diff
@@ -0,0 +1,370 @@
+		out = append(out, fieldName)
+	}
+	sort.Strings(out)
+	return out
+}
+
+func (msg *protoMessage) SetField(name string, val starlark.Value) error {
+	fieldDesc := getFieldDescriptor(msg.msgDesc, name)
+	if fieldDesc == nil {
+		return fmt.Errorf("AttributeError: `%s' value has no field %q", msg.Type(), name)
+	}
+
+	if err := msg.CheckMutable("set field of"); err != nil {
+		return err
+	}
+
+	// Convert starlark.List and starlark.Dict on assignment
+	if fieldDesc.IsList() {
+		if starlarkListVal, ok := val.(*starlark.List); ok {
+			// Convert starlark.List to protoRepeated
+			list, err := newProtoRepeatedFromList(fieldDesc, starlarkListVal)
+			if err != nil {
+				return err
+			}
+
+			val = list
+		}
+	} else if fieldDesc.IsMap() {
+		if starlarkDictVal, ok := val.(*starlark.Dict); ok {
+			// Convert stalark.Map into protoMap
+			mapVal, err := newProtoMapFromDict(fieldDesc.MapKey(), fieldDesc.MapValue(), starlarkDictVal)
+			if err != nil {
+				return err
+			}
+
+			val = mapVal
+		}
+	}
+
+	// Allow using msg_field = None to unset a scalar message field
+	if fieldDesc.Kind() == protoreflect.MessageKind && val == starlark.None && !fieldDesc.IsList() && !fieldDesc.IsMap() {
+		delete(msg.fields, name)
+		return nil
+	}
+
+	// If valueFromStarlark returns an error, the val cannot be assigned to the field
+	_, err := valueFromStarlark(msg.msg.ProtoReflect(), fieldDesc, val)
+	if err != nil {
+		return err
+	}
+
+	// Clear other oneof
+	oneof := fieldDesc.ContainingOneof()
+	if oneof != nil {
+		fields := oneof.Fields()
+		for i := 0; i < fields.Len(); i++ {
+			delete(msg.fields, string(fields.Get(i).Name()))
+		}
+	}
+
+	msg.fields[name] = val
+
+	return nil
+}
+
+// Applies default values for proto2 fields with defaults that are not already set
+func (msg *protoMessage) SetDefaults() error {
+	if err := msg.CheckMutable("set field defaults of"); err != nil {
+		return err
+	}
+
+	for _, fieldName := range fieldNames(msg.msgDesc) {
+		fieldDesc := getFieldDescriptor(msg.msgDesc, fieldName)
+		if fieldDesc == nil {
+			continue
+		}
+
+		// Already set, do not set default
+		if _, ok := msg.fields[fieldName]; ok {
+			continue
+		}
+
+		if !fieldDesc.HasDefault() {
+			continue
+		}
+
+		val, err := valueToStarlark(fieldDesc.Default(), fieldDesc)
+		if err != nil {
+			return err
+		}
+
+		msg.SetField(fieldName, val)
+	}
+
+	return nil
+}
+
+// Merges values from other into msg following proto.Merge logic
+func (msg *protoMessage) Merge(other *protoMessage) error {
+	if msg.Type() != other.Type() {
+		return fmt.Errorf("Cannot merge protobufs of different types: Merge(%s, %s) ", msg.Type(), other.Type())
+	}
+
+	if err := msg.CheckMutable("merge"); err != nil {
+		return err
+	}
+
+	for fieldName, val := range other.fields {
+		fieldDesc := getFieldDescriptor(msg.msgDesc, fieldName)
+		if fieldDesc == nil {
+			continue
+		}
+
+		merged, err := mergeField(msg.fields[fieldName], val)
+		if err != nil {
```

Comment:
same question here about whether it's safe/expected to leave the proto in a partially-merged state